### PR TITLE
1376/david locked keystore

### LINKED ISSUE #1376: Workaround for locked keystore
The keystore in gaia gets locked by the process while in use. On reconnection we stop the gaia process and restart a new process. In some cases, the keystore is not released yet when we start the new gaia process. In that case the process crashes with following message:
```
gaiacli: panic: resource temporarily unavailable
````

Let's create a simple workaround for handling this case.

Proposed workaround:
- detect `resource temporarily unavailable` error
- wait a second
- retry connecting

### PR DESCRIPTION
**Pull request by NodeGuy**
Closes #1376 

Description:

Currently Gaia Lite reports the error `gaiacli: panic: resource temporarily unavailable` if two instances are run at the same time.  This PR moves the error upstream into Voyager as an assertion in `startLCD` to make it easier to diagnose bugs (and has already caught three).

### CODE REVIEW

File: app/src/main/index.js [hunk 1]
```diff
@@ -308,10 +336,14 @@ function stopLCD() {
     try {
       // prevent the exit to signal bad termination warnings
       lcdProcess.removeAllListeners(`exit`)
-      lcdProcess.on(`exit`, resolve)
+
+      lcdProcess.on(`exit`, () => {
+        lcdProcess = null
```
Comment (jbibla) [inline app/src/main/index.js]: what is the thinking re: `null` instead of `false`?

File: app/src/main/index.js [hunk 2]
```diff
@@ -583,14 +615,15 @@ async function pickAndConnect(addressbook) {
       err
     )
     addressbook.flagNodeIncompatible(nodeIP)
+    await stopLCD()
     return await pickAndConnect(addressbook)
```
Comment (jbibla) [inline app/src/main/index.js]: i know this is unrelated to your PR @NodeGuy but why are we still "picking and connecting" CC @faboweb ?

File: test/unit/specs/main.spec.js [hunk 3]
```diff
@@ -575,22 +576,18 @@ describe(`Startup Process`, () => {
...
-      })
+
+      childProcess.spawn = () =>
+        Object.assign(mockSpawnReturnValue(), {
+          stderr: {
+            on: (type, cb) => {
+              // type is always 'data'
+              errorCB = cb
+            },
+            pipe: () => {}
```
Comment (jbibla) [inline test/unit/specs/main.spec.js]: what's the use?

**Comment by jbibla** [comment]
cool PR!

File: app/src/main/index.js (hunk 1 above)
Comment (faboweb) [reply app/src/main/index.js]: false would mix different types for one variable (Promise, Boolean). in JS everything is possible. but it's better to create predictable code I think.

File: app/src/main/index.js (hunk 2 above)
Comment (faboweb) [reply app/src/main/index.js]: can be renamed/refactored imo. feel free to open an issue.

File: test/unit/specs/main.spec.js [hunk 4]
```diff
@@ -508,8 +511,10 @@ describe(`Startup Process`, () => {
       expect(send.mock.calls[0][1]).toBeTruthy() // TODO fix seeds so we can test nodeIP output
     })
 
-    it(`should try another node if user disapproved the hash`, async () => {
-      main.shutdown()
+    // Skip for now because this test takes 2 seconds and we're not making use
+    // of it yet.
+    it.skip(`should try another node if user disapproved the hash`, async () => {
```
Comment (faboweb) [inline test/unit/specs/main.spec.js]: remove it ;)

File: app/src/main/index.js (hunk 1 above)
Comment (jbibla) [reply app/src/main/index.js]: if i understand you correctly... you are saying `null` and `true` are required - adding `false` to the mix would be unnecessary complication?

File: app/src/main/index.js (hunk 2 above)
Comment (jbibla) [reply app/src/main/index.js]: ahh! your PR isn't in yet!

File: app/src/main/index.js (hunk 1 above)
Comment (NodeGuy) [reply app/src/main/index.js]: `false` is traditionally used for boolean propositions (e.g., "There is an `lcdProcess` running.").

`null` is used to communicate absence of value ("`lcdProcess` contains an instance of `ChildProcess` and in this case there isn't one.")

File: app/src/main/index.js [hunk 5]
```diff
@@ -257,6 +258,12 @@ app.on(`ready`, () => createWindow())
 
 // start lcd REST API
 async function startLCD(home, nodeIP) {
+  assert.equal(
+    lcdProcess,
+    null,
+    `Can't start Gaia Lite because it's already running.  Call StopLCD first.`
```
Comment (fedekunze) [inline app/src/main/index.js]: how do I call `StopLCD` ? though the Command line ? If so we'd need to add it to the README or add a bit more detail to the message.

 Also it should be called `StopGaiaLite` instead of `StopLCD`

Comment (NodeGuy) [reply app/src/main/index.js]: Sorry for the confusion; that error message is intended for developers.  It raises a good question:  Should we have a separate channel for developer errors vs. user errors?  Should the user error for this situation show something like, "An error has occurred.  Please restart Voyager and try again."

@faboweb Do we have any precedent for this?

Comment (jbibla) [reply app/src/main/index.js]: if it appears in a modal, it's a user error and should: 
1. be in plain english
2. provide an action for the user to move forward

if it appears in the console, it's for developers and should:
1. be in more detailed plain english
2. provide an action for the developer to move forward

:trollface:  🌮 👍

Comment (NodeGuy) [reply app/src/main/index.js]: I think it already satisfies the second case.

Comment (faboweb) [reply app/src/main/index.js]: @fedekunze does this satisfy your request for change?

Comment (fedekunze) [reply app/src/main/index.js]: ok resolving then

**Comment by fedekunze** [comment]
@NodeGuy fix conflicts and we should be good to merge
